### ISSUE TITLE
node label missing in node_cpu_seconds_total metric? 

### ISSUE THREAD

**Issue by splashx** [coreos/prometheus-operator#2289]
**What did you do?**

Opened Grafana and clicked on ` K8s / USE Method / Node` dashboard 

**What did you expect to see?**

The stats for nodes

**What did you see instead? Under which circumstances?**

`No data points`

* Prometheus Operator version:

image tag: `v0.25`, project version `v0.26`

* Kubernetes version information:

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-05-12T04:12:12Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
```

* Kubernetes cluster kind: 

This is a custom install based on former kubespray

* Manifests:

Default manifests for `v0.26`

* Extra info:

The node-exporter metrics scraping is healthy:

![screenshot 2019-01-11 at 14 23 25](https://user-images.githubusercontent.com/1154133/51036163-8adfed00-15ac-11e9-9944-86ec92b49e8b.png)

Upon further analysis of the Grafana query corresponding to the board `CPU Utilisation`:
 
```
node:node_cpu_utilisation:avg1m{node="$node"}
```

I've checked the expression corresponding to `node:node_cpu_utilisation:avg1m` and [it says](https://github.com/coreos/prometheus-operator/blob/e3462bd72b4b25234dfcd5c4a83c113bcf26ee39/contrib/kube-prometheus/manifests/prometheus-rules.yaml#L133-L138):

```
    - expr: |
        1 - avg by (node) (
          rate(node_cpu_seconds_total{job="node-exporter",mode="idle"}[1m])
        * on (namespace, pod) group_left(node)
          node_namespace_pod:kube_pod_info:)
      record: node:node_cpu_utilisation:av
```

Although in my manifest this looks slightly different, this expression _as is_, can't work because `group_left(node)` requires a label `node` in `node_cpu_seconds_total` and this metric doesn't contain such label:

![screenshot 2019-01-11 at 14 30 14](https://user-images.githubusercontent.com/1154133/51036454-76e8bb00-15ad-11e9-87fb-ec0b475915f3.png)

**Comment by metalmatze** [comment]
Strange. Probably some version mismatches and thus missing labels.

The `node_cpu_seconds_total` metric is matched with the `node_namespace_pod:kube_pod_info:` on `(namespace, pod)` and the node label is from the latter metric.

Does the metric, it gets merged, have the labels?
```
max by(node, namespace, pod) (label_replace(kube_pod_info{job="kube-state-metrics"}, "pod", "$1", "pod", "(.*)"))
```


Can you check your node-exporter version by running `node_exporter_build_info` and the kube-state-metrics version by looking at the deployment?

**Comment by splashx** [comment]
This query:
```
max by(node, namespace, pod) (label_replace(kube_pod_info{job="kube-state-metrics"}, "pod", "$1", "pod", "(.*)"))
```
returns (34 items in my case, showing here the first 3):

<img width="759" alt="screenshot 2019-01-11 at 19 26 51" src="https://user-images.githubusercontent.com/1154133/51052471-09518480-15d7-11e9-85be-e3bca710bda0.png">

> Can you check your node-exporter version by running node_exporter_build_info 

The node-exporter version is `0.16.0` (revision `d42bd70f4363dced6b77d8fc311ea57b63387e4f`).


> and the kube-state-metrics version by looking at the deployment?

The deployment is using `quay.io/coreos/kube-state-metrics:v1.4.0`.

**Comment by splashx** [comment]
It's important to note that other Dashboards (`K8s / Compute Resources / {Pod, Cluster, Namespace}`, `Pods` and `Nodes`) are working just fine...

**Comment by splashx** [comment]
After all I think the issue is `node_cpu_seconds_total`'s label `instance` has value in the format of IP but `node_namespace_pod:kube_pod_info:`'s label `node` is using FQDN. So `group_left` can't find the common pattern 😕

**Comment by metalmatze** [comment]
I wonder why that is. I guess, it's something that comes from your Kubernetes as `kube_pod_info` only re-maps Kubernetes objects.

**Comment by metalmatze** [comment]
When you pick a random Pod and do `kubectl describe` on it, what do you see as Node value?
Example for my kube-system/kube-apiserver:
```
Node:               10.135.73.45/10.135.73.45
```

And then if you do `kubectl get nodes`?
```
10.135.XXX.194    Ready    node                125d   v1.11
10.135.XXX.142    Ready    node                128d   v1.11
10.135.XXX.45     Ready    controller,master   128d   v1.11
10.135.XXX.73     Ready    node                128d   v1.11
```

**Comment by brancz** [comment]
I think there is a misunderstanding of `group_left`. The `on` statement is what defines what is being joined on, whereas the `group_left` statement says, that the result of the join will be the value and labels on the left side, plus the `node` label of the right side.

**Comment by metalmatze** [comment]
You're totally right, thanks for pointing out.
However the values for the name label should be without the port for the metric to make sense, correct?

**Comment by brancz** [comment]
Yes but as I read it @splashx only said the `instance` label has IP+port, which would be correct, and the `kube_pod_info` metric has the `node` label with the full node name, no? (if so, then all seems correct)

What doesn't seem right is that the node-exporter targets have no `pod` label, that seems wrong.

**Comment by simonmeggle** [comment]
Just stumbled upon this issue because I was searching for an answer/reason for this statement: 

    max by(node, namespace, pod) (label_replace(kube_pod_info{job="kube-state-metrics"}, "pod", "$1", "pod", "(.*)"))

Just for curiosity: why replacing a label value with the same value again? Wouldn't this be shorter and the same? 

    max(kube_pod_info{job="kube-state-metrics"}) by (node, namespace, pod) 

(PS: Sorry, don't want to hijack this issue...)

**Comment by brancz** [comment]
The use `label_replace` function is indeed not necessary in this setup, it ends up here because we use the [kubernetes-mixin](https://github.com/kubernetes-monitoring/kubernetes-mixin) to generate this rule, and the mixin allows configuring what the `pod` label is in any particular setup, it just happens, that in our setup the label for a Pod is `pod`.

**Comment by simonmeggle** [comment]
Ok, thanks for the clarification!

**Comment by konvergence** [comment]
Whhy did you remove node label ?
